HyperPlay-Gaming/hyperplay-desktop-client
# [UX] Improve Download Toast

WIP

---

Use the following Checklist if you have changed something on the Backend or Frontend:


---

## 🐛 Issue Addressed

**Linked Issue:** [Install modal does not display extracting](https://github.com/HyperPlay-Gaming/hyperplay-desktop-client/issues/455) - PR Main issue
**Linked Issue:** [Disk space check does not account for extraction space](https://github.com/HyperPlay-Gaming/hyperplay-desktop-client/issues/614)

---

## ✅  Other Fixed Issues

-  Progress on Download does not start from 0, generally was started either from previous queued progress if it was canceled earlier
- Disk Space is another issue mentioned by @nyghtstalker, and fixed with suggest of @flavioislima. Currently, we are utilizing a basic function to estimate disk space. Typically, the estimated size is doubled and then increased by 0.05-0.01%, depending on whether it is being compressed on a Mac (higher percentage) or on Windows and other operating systems.
- Upon relaunch the download modal appears again even though there is nothing downloading. This should be fixed, although @nyghtstalker, please make sure to check it out as soon this PR read for review

---

## 🚀 Props API Update for Components

- Updated GameState to handle the new status of `inExtraction`
- Updated `DownloadToast` with a new `downloadStatus` type: 'inExtraction'

## ☕ New PCI Handlers

- cancelExtraction | window.api.cancelExtraction(appName: string). This method allows you to cancel an extraction, generally would require an extraction in progress.

---

## 📌 PR Blocked From Issue Below

[HyperPlay-Gaming/hyperplay-ui#101](https://github.com/HyperPlay-Gaming/hyperplay-ui/pull/101)

---

## 🗺️ Approach Taken

Our aim is to start using classes because we have a lot of code that is intertwined and makes debugging, fixing, or enhancing the code time-consuming. For instance, we have developed an `ExtractZipService` class that handles zip extraction-related functionalities, including the decompressed total bytes for later use. Our approach enables us to test the class independently, providing us with a better guarantee that everything is working as expected. Additionally, we always apply the SOC Principle, which helps us avoid days of debugging. 🙌 

---

## 🛠️ Testing Steps

Before diving in, ensure you have access to both `hyperplay-ui` and `hyplerplay-store` repositories.

1. Navigate to the feature branch:
```bash
git fetch && git checkout 455-update-game-card-and-download-toast-for-extraction && git pull
```
2. If you haven't already, clone the `hyperplay-ui`:
```bash
git clone https://github.com/HyperPlay-Gaming/hyperplay-ui.git
```
3. Switch to the relevant branch in `hyperplay-ui`:
```bash
git fetch && git 455-update-game-card-and-download-toast-for-extraction && git pull
```
4. Link `hyperplay-ui` to `hyperplay-desktop-client`:
```bash
yarn add path/to/your/hyperplay-ui
```
5. Fire up the development server:
```bash
yarn start
```
6. Next head to Store > Find `Megaweapon` > Add to Library
7. Next head to your library
8. Tap Download Icon
9. Install Model will show, which generally you would be already able to see the uncompressed size (`estimated`)
10. Tap Install
11. Wait until the extraction process
12. Wait until it does install

---

### ⚠️ Review Tests

Please make sure to test all the following, please feel free to edit the PR readme in case there are more ways we should test

**Pages**
- [ ] Check on Library
- [ ] Check on game page
- [ ] Check on the Download Manager page
- [ ] I don't know how we can test it yet, but try to install it through the store

**Toast, Download Bar on Game Card, Game Actions on Download Manager, Buttons on Game Page (Cancel Extraction, Cancel Download, etc)** 
- [ ] Pause Download
- [ ] Pause Download & Cancel
- [ ] Pause Download & Cancel & Reinitiate 
- [ ] Resume Download
- [ ] Resume Download & Cancel
- [ ] Resume Download & Cancel & Reinitiate the download 
- [ ] Cancel Extraction 
- [ ] Cancel Extraction & Reinitiate the download

**UI**
- [ ] Download Toast must have Pause or Resume Button, Cancel Button
- [ ] Download Bar Game Card must have Pause or Resume Button, Cancel Button
- [ ] Extraction Toast must only Cancel Button
- [ ] Extraction Bar Game Card must have Pause or Resume Button, Cancel Button
- [ ] Extraction Game Page must have Cancel Download Button
- [ ] Extraction Game Page must only Cancel Extraction Button

---

## 📣 Important Note

This fix pinpoints where pause, resume are generally controlled by a status on `hyperplay-ui` lib, which we had to add. Once we're all aligned, let's ensure to 
- We merge [HyperPlay-Gaming/hyperplay-ui#101](https://github.com/HyperPlay-Gaming/hyperplay-ui/pull/101). 
- Quick update to the `package.json` of `hyperplay-ui` will be needed, with the latest update
- Without the merge of [HyperPlay-Gaming/hyperplay-ui#101](https://github.com/HyperPlay-Gaming/hyperplay-ui/pull/101) then the Code Check will fail, due of the interface.
